### PR TITLE
Intercept separation

### PR DESCRIPTION
**Pull request by biyan-linkedin**
Implement intercept separation which will exclude penalizing intercept in l2 regularization.
 
Add UT and integTest for this feature.

### CODE REVIEW

File: .gitattributes [hunk 1]
```diff
@@ -1,4 +1,2 @@
-photon-api/src/integTest/resources/* linguist-vendored
-photon-api/src/test/resources/* linguist-vendored
-photon-client/src/integTest/resources/* linguist-vendored
-photon-client/src/test/resources/* linguist-vendored
+photon-ml/src/integTest/resources/* linguist-vendored
+photon-ml/src/test/resources/* linguist-vendored
```
Comment (ashelkovnykov) [inline .gitattributes]: What's the purpose of this change?

File: .travis.yml [hunk 2]
```diff
@@ -6,7 +6,6 @@ scala:
 sudo: required
 jdk:
   - oraclejdk8
-dist: trusty
```
Comment (ashelkovnykov) [inline .travis.yml]: What's the purpose of this change?

File: photon-all/build.gradle [hunk 3]
```diff
@@ -16,9 +16,7 @@ apply plugin: 'scala'
 apply plugin: 'com.github.johnrengelman.shadow' // See doc at: http://imperceptiblethoughts.com/shadow/.
 
 dependencies {
-  compile project(":photon-lib$scalaSuffix")
-  compile project(":photon-api$scalaSuffix")
-  compile project(":photon-client$scalaSuffix")
+  compile project(":photon-ml$scalaSuffix")
```
Comment (ashelkovnykov) [inline photon-all/build.gradle]: What's the purpose of this change?

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala [hunk 4]
```diff
@@ -186,28 +187,30 @@ object RandomEffectCoordinate {
    *                               problems
    * @param randomEffectDataset The data on which to run the optimization algorithm
    * @param configuration The optimization problem configuration
-   * @param objectiveFunction The objective function to optimize
+   * @param objectiveFunctionFactory The objective function to optimize
    * @param glmConstructor The function to use for producing GLMs from trained coefficients
    * @param normalizationContext The normalization context
    * @param varianceComputationType If and how coefficient variances should be computed
-   * @return A new [[RandomEffectCoordinate]] object
+   * @return
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala]: The return type is still the same

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala [hunk 5]
```diff
@@ -229,10 +232,11 @@ object RandomEffectCoordinate {
   private def buildRandomEffectOptimizationProblem[RandomEffectObjective <: SingleNodeObjectiveFunction](
       linearSubspaceProjectorsRDD: RDD[(REId, LinearSubspaceProjector)],
       configuration: RandomEffectOptimizationConfiguration,
-      objectiveFunction: RandomEffectObjective,
+      objectiveFunctionFactory: SingleNodeObjectiveFunctionFactory,
       glmConstructor: Coefficients => GeneralizedLinearModel,
       normalizationContext: NormalizationContext,
-      varianceComputationType: VarianceComputationType = VarianceComputationType.NONE): RandomEffectOptimizationProblem[RandomEffectObjective] = {
+      varianceComputationType: VarianceComputationType = VarianceComputationType.NONE,
+      interceptIndexOpt: Option[Int]): RandomEffectOptimizationProblem[SingleNodeObjectiveFunction] = {
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala]: Why change the return type?

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala [hunk 6]
```diff
@@ -131,6 +131,12 @@ class GameEstimator(val sc: SparkContext, implicit val logger: Logger) extends P
     "use warm start",
     "Whether to train the current model with coefficients initialized by the previous model.")
 
+  val coordinateInterceptIndices: Param[Map[CoordinateId, Int]] =
+    ParamUtils.createParam[Map[CoordinateId, Int]](
+    "coordinate intercept indices",
+    "A map of coordinate ID to intercept index, if enabled."
+  )
```
Comment (ashelkovnykov) [inline photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: Please move closing bracket onto the line above (137). Please also drop "if enabled" from the parameter description (it doesn't make sense, since there is a default value for it, thus it is always enabled).

File: photon-api/src/test/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblemTest.scala [hunk 7]
```diff
@@ -90,10 +90,10 @@ class DistributedOptimizationProblemTest {
...
   @Test(dataProvider = "varianceInput")
   def testComputeVariances(
-      varianceComputationType: VarianceComputationType,
-      optimizer: Optimizer[DistributedObjectiveFunction],
-      objectiveFunction: DistributedObjectiveFunction,
-      expected: Option[Vector[Double]]): Unit = {
+    varianceComputationType: VarianceComputationType,
+    optimizer: Optimizer[DistributedObjectiveFunction],
+    objectiveFunction: DistributedObjectiveFunction,
+    expected: Option[Vector[Double]]): Unit = {
```
Comment (ashelkovnykov) [inline photon-api/src/test/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblemTest.scala]: Previous style was correct

File: photon-api/src/test/scala/com/linkedin/photon/ml/optimization/GeneralizedLinearOptimizationProblemTest.scala [hunk 8]
```diff
@@ -202,10 +202,10 @@ object GeneralizedLinearOptimizationProblemTest {
...
 
   private class MockOptimizationProblem(
-      optimizer: Optimizer[SingleNodeSmoothedHingeLossFunction],
-      objectiveFunction: SingleNodeSmoothedHingeLossFunction,
-      glmConstructor: Coefficients => GeneralizedLinearModel,
-      regularizationContext: RegularizationContext)
+    optimizer: Optimizer[SingleNodeSmoothedHingeLossFunction],
+    objectiveFunction: SingleNodeSmoothedHingeLossFunction,
+    glmConstructor: Coefficients => GeneralizedLinearModel,
+    regularizationContext: RegularizationContext)
```
Comment (ashelkovnykov) [inline photon-api/src/test/scala/com/linkedin/photon/ml/optimization/GeneralizedLinearOptimizationProblemTest.scala]: Previous style was correct

File: photon-api/src/test/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblemTest.scala [hunk 9]
```diff
@@ -85,12 +85,12 @@ class SingleNodeOptimizationProblemTest {
...
+@Test(dataProvider = "varianceInput")
   def testComputeVariances(
-      varianceComputationType: VarianceComputationType,
-      optimizer: Optimizer[SingleNodeObjectiveFunction],
-      objectiveFunction: SingleNodeObjectiveFunction,
-      expected: Option[Vector[Double]]): Unit = {
+    varianceComputationType: VarianceComputationType,
+    optimizer: Optimizer[SingleNodeObjectiveFunction],
+    objectiveFunction: SingleNodeObjectiveFunction,
+    expected: Option[Vector[Double]]): Unit = {
```
Comment (ashelkovnykov) [inline photon-api/src/test/scala/com/linkedin/photon/ml/optimization/SingleNodeOptimizationProblemTest.scala]: Previous style was correct

File: photon-api/src/test/scala/com/linkedin/photon/ml/util/GameTestUtils.scala [hunk 10]
```diff
@@ -296,12 +296,12 @@ trait GameTestUtils extends SparkTestUtils {
...
-      numEntities: Int,
-      size: Int,
-      dimensions: Int,
-      seed: Int = DefaultSeed): (RandomEffectCoordinate[SingleNodeGLMLossFunction], RandomEffectModel) = {
+    randomEffectType: REType,
+    featureShardId: FeatureShardId,
+    numEntities: Int,
+    size: Int,
+    dimensions: Int,
+    seed: Int = DefaultSeed): (RandomEffectCoordinate[SingleNodeGLMLossFunction], RandomEffectModel) = {
```
Comment (ashelkovnykov) [inline photon-api/src/test/scala/com/linkedin/photon/ml/util/GameTestUtils.scala]: Previous style was correct

File: photon-client/build.gradle [hunk 11]
```diff
@@ -17,8 +17,7 @@ apply from: '../build-scripts/integration-test.gradle'
 
 dependencies {
   compile(project(':photon-avro-schemas'))
-  compile(project(":photon-lib$scalaSuffix"))
-  compile(project(":photon-api$scalaSuffix"))
+  compile(project(":photon-core$scalaSuffix"))
```
Comment (ashelkovnykov) [inline photon-client/build.gradle]: I think you have an issue with your branch history, because you're undoing older commits in this one

File: .gitattributes (hunk 1 above)
Comment (biyan-linkedin) [reply .gitattributes]: It was generated by Intelij. Will remove

File: .travis.yml (hunk 2 above)
Comment (biyan-linkedin) [reply .travis.yml]: It was generated by Intelij. Will remove

File: photon-all/build.gradle (hunk 3 above)
Comment (biyan-linkedin) [reply photon-all/build.gradle]: It was generated by Intelij. Will remove

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala (hunk 5 above)
Comment (biyan-linkedin) [reply photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/RandomEffectCoordinate.scala]: It does not compile.

**Comment by li-ashelkov** [comment]
Closing PR at request of @biyan-linkedin  - see PR #440.
